plutus-core/plutus-ir/src/PlutusIR/Analysis/RetainedSize.hs

```diff
@@ -0,0 +1,215 @@
+bindingSize
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Binding tyname name uni fun ann -> UniqueMap Unique Size
+bindingSize (TermBind _ _ var term) =
+    insertByNameIndex var (varDeclSize var <> termSize term) mempty
+bindingSize (TypeBind _ tyVar ty) =
+    insertByNameIndex tyVar (tyVarDeclSize tyVar <> typeSize ty) mempty
+bindingSize (DatatypeBind _ (Datatype _ dataDecl params matchName constrs))
+    = insertByNameIndex dataDecl (tyVarDeclSize dataDecl)
+    . flip (foldr $ \param -> insertByNameIndex param $ tyVarDeclSize param) params
+    . insertByNameIndex matchName (Size 1)
+    . flip (foldr $ \constr -> insertByNameIndex constr $ varDeclSize constr) constrs
+    $ mempty
+
+-- | Construct a 'UniqueMap' having size information for each individual part of every 'Binding'
+-- in a term.
+bindingSizes
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Term tyname name uni fun ann -> UniqueMap Unique Size
+bindingSizes (Let _ _ binds term) = foldMap bindingSize binds <> bindingSizes term
+bindingSizes term                 = term ^. termSubterms . to bindingSizes
+
+-- | Same as 'bindingSizes' but is wrapped in a newtype and has a bogus entry for the root.
+toDirectionRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique)
+    => Term tyname name uni fun ann -> DirectionRetentionMap
+toDirectionRetentionMap term =
+    DirectionRetentionMap . IntMap.insert rootInt rootSize . unUniqueMap $ bindingSizes term where
+        -- See Note [Handling the root].
+        rootSize = Size (- 10 ^ (10::Int))
+
+-- | Check if a 'Node' appears in 'DirectionRetentionMap'.
+hasSizeIn :: DirectionRetentionMap -> Node -> Bool
+hasSizeIn _             Root                                   = True
+hasSizeIn (DirectionRetentionMap ss) (Variable (PLC.Unique i)) = i `IntMap.member` ss
+
+-- | Compute the retention map of a term.
+termRetentionMap
+    :: (HasUnique tyname TypeUnique, HasUnique name TermUnique, ToBuiltinMeaning uni fun)
+    => Term tyname name uni fun ann -> IntMap Size
+termRetentionMap term = depsRetentionMap sizeInfo deps where
+    sizeInfo = toDirectionRetentionMap term
+    deps = C.induce (hasSizeIn sizeInfo) $ fst $ runTermDeps term
+
+-- | Apply a function to the annotation of each part of every 'Binding' in a term.
+reannotateBindings
+    :: (HasUnique name TermUnique, HasUnique tyname TypeUnique)
+    => (Unique -> ann -> ann)
+    -> Term tyname name uni fun ann
+    -> Term tyname name uni fun ann
+reannotateBindings f = goLet where
+    -- We don't need these helper functions anywhere else, so we make them into local definitions.
+    goVarDecl (VarDecl ann name ty) = VarDecl (f (name ^. theUnique) ann) name ty
+    goTyVarDecl (TyVarDecl ann tyname kind) = TyVarDecl (f (tyname ^. theUnique) ann) tyname kind
+    goDatatype (Datatype ann dataTyDecl paramTyDecls matchName constrDecls) =
+        Datatype
+            (f (matchName ^. theUnique) ann)
```

Comment:
odd choice? worth a comment

Reply:
It's not odd if it's the only option :)
Plus, we're lucky that the matcher is the only thing that survives erasure, so it's not even that odd.

We do the same for scoping analysis.

I agree it's peculiar enough for us to be great to have a separate step putting uniques into annotations (plus some additional info needed for scoping: is it a binding, is it a variable etc) that can be reused both here and for scoping.